packages/run-protocol/src/vaultFactory/vaultDirector.js

```diff
@@ -0,0 +1,333 @@
+
+  const collateralTypes = makeScalarMap('brand');
+
+  const vaultParamManagers = makeScalarMap('brand');
+
+  return {
+    collateralTypes,
+    debtMint,
+    electorateParamManager,
+    mintSeat,
+    rewardPoolSeat,
+    penaltyPoolSeat,
+    vaultParamManagers,
+    zcf,
+  };
+};
+
+/**
+ * "Director" of the vault factory, overseeing "vault managers".
+ *
+ * @param {ZCF<GovernanceTerms<{}> & {
+ *   ammPublicFacet: AutoswapPublicFacet,
+ *   liquidationInstall: Installation<import('./liquidateMinimum.js').start>,
+ *   loanTimingParams: {ChargingPeriod: ParamRecord<'nat'>, RecordingPeriod: ParamRecord<'nat'>},
+ *   timerService: TimerService,
+ *   priceAuthority: ERef<PriceAuthority>
+ * }>} zcf
+ * @param {import('@agoric/governance/src/contractGovernance/typedParamManager').TypedParamManager<{Electorate: "invitation"}>} electorateParamManager
+ * @param {ZCFMint<"nat">} debtMint
+ */
+const makeVaultDirector = (zcf, electorateParamManager, debtMint) => {
+  const {
+    collateralTypes,
+    // electorateParamManager,
+    mintSeat,
+    rewardPoolSeat,
+    penaltyPoolSeat,
+    vaultParamManagers,
+    // zcf,
+  } = initState(zcf, electorateParamManager, debtMint);
+
+  /** Make a loan in the vaultManager based on the collateral type.
+   *
+   * @deprecated
+   */
+  const makeVaultInvitation = () => {
+    /** @param {ZCFSeat} seat */
+    const makeVaultHook = async seat => {
+      assertProposalShape(seat, {
+        give: { Collateral: null },
+        want: { RUN: null },
+      });
+      const {
+        give: { Collateral: collateralAmount },
+      } = seat.getProposal();
+      const { brand: brandIn } = collateralAmount;
+      assert(
+        collateralTypes.has(brandIn),
+        X`Not a supported collateral type ${brandIn}`,
+      );
+      /** @type {VaultManager} */
+      const mgr = collateralTypes.get(brandIn);
+      return mgr.makeVaultKit(seat);
+    };
+    return zcf.makeInvitation(makeVaultHook, 'MakeVault');
+  };
+
+  // TODO put on machineFacet for use again in publicFacet
+  const getCollaterals = async () => {
+    // should be collateralTypes.map((vm, brand) => ({
+    return harden(
+      Promise.all(
+        [...collateralTypes.entries()].map(async ([brand, vm]) => {
+          const priceQuote = await vm.getCollateralQuote();
+          return {
+            brand,
+            interestRate: vm.getGovernedParams().getInterestRate(),
+            liquidationMargin: vm.getGovernedParams().getLiquidationMargin(),
+            stabilityFee: vm.getGovernedParams().getLoanFee(),
+            marketPrice: makeRatioFromAmounts(
+              getAmountOut(priceQuote),
+              getAmountIn(priceQuote),
+            ),
+          };
+        }),
+      ),
+    );
+  };
+
+  /** @type {VaultFactory} */
+  const machineBehavior = {
+    // TODO move this under governance #3972
+    /** @type {AddVaultType} */
+    addVaultType: async (
```

Comment:
`addVaultType` is too big to be inline in `machineBehavior`. It really ought to be top-level.  Since it has internal `await`s, it has to be top-level anyway.

Reply:
This is en route to `defineKindMulti` where it will have one level less of nesting. This just made it easier to move.

> Since it has internal awaits, it has to be top-level anyway.

Pretty sure that nesting rule is only about `awaits` not the function definition location. https://github.com/Agoric/agoric-sdk/wiki/No-Nested-Await